### ISSUE TITLE
Feature request: Adding timing for text

### ISSUE THREAD

**Issue by iuqiddis** [lettier/gifcurry#22]
Hi again. Would you consider adding a way to add timing information for overlay text? For example if I'm making a gif of a single back-and-forth conversation, it would be nice to have the first line displayed for _x_ seconds, then the second line is displayed for _y_ seconds. 

I was thinking a possible way to implement it in the gui, perhaps something like the attached file.

![with_timing_anno](https://user-images.githubusercontent.com/13721585/39165855-7bcd7714-4754-11e8-99ba-ed9e824a1e2d.png)

Labels 1 and 2 point to the fields --similar to how you implemented it for selecting clip start and duration--  for the displayed text. 

Label 3, which points to the plus button, is saying if you want to add a second line of text with different timing. Perhaps hitting the plus button adds a similar field right below it, or pops up a dialogue, like for selecting fonts. I could mock that up too, if what I'm saying is not clear.

Maybe this adds too much complexity to the program and is beyond the intended scope, but if not, it would be a cool feature to add. Thanks again, lovely program to use.

**Comment by lettier** [comment]
Hello @iuqiddis,

Thank you for the feature request. 

Multiple time based text overlays are currently in the works. Below is a mock up demonstration.

![gifcurry](https://user-images.githubusercontent.com/1661343/39225344-cfc2e56e-4819-11e8-8dfd-d1fd2a17e2bd.gif)

You'll be able to click Add and it will display another text overlay. For each text overlay, you can set the font, size, left/right/up/down position, and the start and duration time.

This feature and being able to see the text in the preview are slated for the next major release.

I'll update this issue once the feature is finished.

Thank you for the kind words and I'm glad you're enjoying Gifcurry.

:+1:

**Comment by iuqiddis** [comment]
awesome. looks great. 
cheers.

**Comment by lettier** [comment]
Hello @iuqiddis,

Wanted to give you an update regarding the multiple time based text overlays feature.

It is almost complete and I'll be uploading a pre-release AppImage soon.

![test](https://user-images.githubusercontent.com/1661343/41013245-2b6af874-6913-11e8-9169-56fc213ccc2e.gif)

:+1:

**Comment by lettier** [comment]
Hello @iuqiddis,

You can now try out the pre-release AppImage for version [4.0.0.0](https://github.com/lettier/gifcurry/releases/tag/4.0.0.0).

:+1:

**Comment by iuqiddis** [comment]
The feature works great. I tried it on three different clips, easy enough to move timing and placement. thanks a lot!

can i ask about some other features introduced in version 4.0.0.0 in this thread? they are not related to the text overlaying interface. Or should i start a new thread for other comments or requests. thanks again!

**Comment by iuqiddis** [comment]
cool. 

1. (this is is related to the text overlay) Is there a possibility of adding a parameter for outline weight?

2. Previously the output was as .mp4 and the quality seemed a lot better (when going for the higher quality). The webm version, even at the 'insane' quality seems pretty low quality in the clips i've tried. I can reply later with a comparison between version 3 and 4 and upload it back with the source file you shared in a different thread (I assume its not some bug on my computer). BUT, coming to my request, would you consider having output options for both h264 and webm (also because safari doesn't support webm unfortunately)

3. Lastly (so sorry for the length), is it possible to implement importing embedded subtitles with timing and load them as pre-loaded text? This might be an edge case, but a lot of times I cut 2-3second scenes with dialogue (hence the timing for text) and often my sources have embedded subtitles. It would be cool if that can get atomically imported.

cheers.

**Comment by lettier** [comment]
Hello @iuqiddis,

1) Sure. Would this increase and decrease the size of the text outline? I could add it among the current text controls.

2)

It would be great to add it as a configurable option but that just isn't tenable at the moment unfortunately. :disappointed:  Gifcurry will have to stay with the royalty free codec VP9 inside the WebM container for now. Even though Gifcurry does not distribute, implement, and/or link (neither statically or dynamically) with an implementation of any royalty-burdened (or any for that matter) codec, it is safer to stick with the WebM VP9 combination.

That said, the VP9 codec is very comparable to what you might find used in MP4 and in some cases it is better. The good news is I have increased the quality of the output video considerably to that of 1080p HD video. I've also increased the insane quality level by removing all color count restrictions. When saving as a video, Gifcurry now bypasses the GIF creation step and goes straight to creating the video after the annotation step.

Input video: https://svs.gsfc.nasa.gov/vis/a010000/a012000/a012034/Ice_Edge_UHD_YouTube.mp4
Sample output video (900 width @ Insane): https://streamable.com/zg03z
Sample output GIF (900 width @ High): 
![gif-test-high](https://user-images.githubusercontent.com/1661343/41273267-cbf1a66c-6de6-11e8-93b6-64f586bf36fd.gif)

You can try out the new settings using the [pre-release 4.0.0.0 AppImage](https://github.com/lettier/gifcurry/releases/download/4.0.0.0/gifcurry-4.0.0.0-x86_64.AppImage).

In the future, Gifcurry will switch to the [AV1 codec](https://bitmovin.com/av1/) which should be supported by Safari since [Apple joined the consortium](https://9to5mac.com/2018/01/04/apple-alliance-for-open-media/).

For now, you can pass Gifcurry's output to [Curlew](https://www.fossmint.com/curlew-media-converter-for-linux/) or in the terminal just type `ffmpeg -i gifcurry.webm gifcurry.mp4`.

3) I could offer a load option for a [Structured Subtitle Format (SRT)](https://matroska.org/technical/specs/subtitles/srt.html) file that would then automatically create all of the texts. Or are you saying it would read the subtitles directly from the input video file? This may be difficult depending on how many languages are stored with the video. If they are "burned in" then it would not be possible to read them from the video file.

:+1:

**Comment by iuqiddis** [comment]
1. yes, exactly. just decreasing the size of the outline. thanks.

2. I tried the app image link with one of my sources and yup, the insane quality is much nicer now (I'd consider it good/great, whereas the previous output was pretty poor). 

For this particular scene, the previous 4.0.0.0 was outputting both 'insane' and 'low' webm at ~160 kB, which besides being strange, resulted in poor output quality in both cases. Now the 'insane' setting with AppImage download results in a ~1.0MB high quality output which is awesome; for comparison the 'medium' setting GIF is 24MB and a lot poorer in quality (but still usaable i suppose). I'm using 720p resolution as output to play around with the clips, thus the high GIF size.

|Version | Quality | Size (kB) | Qualitative Assessment |
|---|---|---|---|
|new|insane	|1121|high/great|
|new|high	|1121|high/great|
|new|medium	|937 |maybe slightly worse than above two|
|new|low	|833 |slightly worse. still perfectly adaquate|
|old|insane	|158|blocky/not good|
|old|high	|158|blocky/not good|
|old|low	|165|blocky/not good|

For a different clip, I get similar ratios between insane and low (~83%):

|Version | Quality | Size (kB) | Qualitative Assessment |
|---|---|---|---|
|new|insane	|770|high/great|
|new|high	|770|high/great|
|new|low	|636 |slightly worse. still perfectly adaquate|

I was going to suggest maybe playing around with the settings to only have **insane**, **medium** and **low**, where the low setting generates a file that is approximately 50-60% in size of the **insane** output? Based on my meager testing on two action-based clips, the the difference between **insane** and **high** webm doesn't seem to be much (as seen in the above tables).

As for going for webm, yeah, I can understand only having the option for the royalty-free option. I think you're playing it safe, but that's your call :) . I'll just stick to ffmpeg for h264 output, as you suggested.

3. I was thinking more along just loading the embedded subtitles from the source clip to start off with. I assumed this would _exclude_ burned-in subtitles. Maybe default to the first embedded track, or look for a particular language (English in my case, though of course that would vary by the user's wants or needs).

Thanks again for listening. Really appreciate your responses. Cheers.

**Comment by lettier** [comment]
Hello @iuqiddis,

1) I have updated the pre-release AppImage to include an outline size selection. I also added text rotation.

2) I went ahead and tweaked the settings reducing the quality options to only High, Medium, and Low.

Input video: https://svs.gsfc.nasa.gov/vis/a010000/a012000/a012034/Ice_Edge_UHD_YouTube.mp4
- Start second: 60
- Duration seconds: 1

Video output at 1280x720:
- High: 904 kB
- Medium: 246 kB
- [Low: 182 kB](https://streamable.com/lmenr)

3) I could do something similar that I do for [Movie Monad](https://github.com/lettier/movie-monad) where it will list the available languages, you would select one, and then it would import them for you.

:+1:
